src/main/java/se/knowit/bookitregistration/repository/map/RegistrationRepositoryMapImpl.java

```diff
@@ -1,64 +1,55 @@
-import se.knowit.bookitregistration.service.RegistrationService;
+import se.knowit.bookitregistration.repository.RegistrationRepository;
 import se.knowit.bookitregistration.service.exception.ConflictingEntityException;
 
 import java.util.Map;
-import java.util.Optional;
 import java.util.Set;
 import java.util.UUID;
 import java.util.concurrent.ConcurrentHashMap;
 import java.util.concurrent.atomic.AtomicLong;
 import java.util.function.Predicate;
 import java.util.function.Supplier;
-import java.util.stream.Collectors;
 
-public class RegistrationServiceMapImpl implements RegistrationService {
-    private final Map<Long, Registration> registrationStore;
+import static java.util.stream.Collectors.toSet;
+import static java.util.stream.Collectors.toUnmodifiableSet;
+
+public class RegistrationRepositoryMapImpl implements RegistrationRepository {
+
     private IdentityHandler identityHandler;
-    private RegistrationValidator registrationValidator;
+    private final Map<Long, Registration> registrationStore;
 
-    public RegistrationServiceMapImpl() {
+    public RegistrationRepositoryMapImpl() {
         this(new ConcurrentHashMap<>());
     }
 
-    RegistrationServiceMapImpl(Map<Long, Registration> registrationStore) {
+    public RegistrationRepositoryMapImpl(Map<Long, Registration> registrationStore) {
         this.registrationStore = registrationStore;
         this.identityHandler = new IdentityHandler(this.registrationStore.size());
-        this.registrationValidator = new RegistrationValidator();
     }
 
     @Override
-    public Set<Registration> findAll() {
-        return Set.copyOf(registrationStore.values());
-    }
-
-    @Override
-    public Registration save(Registration incomingRegistration) throws ConflictingEntityException {
-        Registration validRegistration = registrationValidator.ensureRegistrationIsValidOrThrowException(incomingRegistration);
-        assignRequiredIds(validRegistration);
-        persistRegistration(validRegistration);
-        return validRegistration;
+    public Set<Registration> find(Predicate<Registration> searchFilter) {
+        return registrationStore.values().stream().filter(searchFilter).collect(toSet());
     }
 
     @Override
-    public void delete(String registrationId) {
-        Optional<Registration> registrationToDelete = registrationStore.values()
+    public void delete(Predicate<Registration> searchFilter) {
+        Set<Long> idsOfRegistrationsToDelete = registrationStore.values()
                 .stream()
-                .filter(r -> r.getRegistrationId().toString().equals(registrationId))
-                .findFirst();
-        registrationToDelete.ifPresent(registration -> registrationStore.remove(registration.getId()));
+                .filter(searchFilter)
+                .map(Registration::getId)
+                .collect(toUnmodifiableSet());
+
+        idsOfRegistrationsToDelete.forEach(registrationStore::remove);
```

Comment:
Can't we add this to the steam?

Reply:
I don't think it's good practice to mutate the source of the stream while working on the stream. Or??

Reply:
What Ola said :)

I'm not even sure the stream API allows you to, no :)

Reply:
With a different predicate it can be done:
registrationStore.entrySet().removeIf(searchFilter);